### PR TITLE
[SPARK-11802][SQL] Kryo-based encoder for opaque types in Datasets

### PR DESCRIPTION
**Pull request by rxin**
I also found a bug with self-joins returning incorrect results in the Dataset API. Two test cases attached and filed SPARK-11803.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/Encoder.scala [hunk 1]
```diff
@@ -37,7 +38,33 @@ trait Encoder[T] extends Serializable {
...
+/**
+ * Methods for creating encoders.
+ */
 object Encoders {
+
+  /**
+   * (Scala-specific) Creates an encoder that serializes objects of type T using Kryo.
+   * This encoder maps T into a single byte array (binary) field.
+   */
+  def kryo[T: ClassTag]: Encoder[T] = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/Encoder.scala]: How about putting this method into `ExpressionEncoder`? I think `Encoders` is only used at java side as the lack of implicit magic.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/Encoder.scala]: then we can change the return type to `ExpressionEncoder[T]` and make the tests less verbose with `asInstansOf`

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/Encoder.scala]: what if we want to use kryo in scala?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/Encoder.scala]: oh, this can't be done by implicit, I was wrong, nvm.

**Pull request closed without merging**